Geometry/CaloEventSetup/plugins/CaloGeometryBuilder.cc

```diff
@@ -58,76 +98,60 @@ CaloGeometryBuilder::produceAligned( const CaloGeometryRecord& iRecord )
-      }
-      else if ( ite->find(HGCalGeometry::producerTag()) != std::string::npos ) {
-	edm::LogInfo("CaloGeometryBuilder") << "Building " << *ite << " reconstruction geometry";
-	edm::ESHandle<HGCalGeometry> pHG;
-	iRecord.getRecord<IdealGeometryRecord>().get(*ite,pHG);
-	const auto& topo = pHG->topology();
-	pCalo->setSubdetGeometry(topo.detector(),topo.subDetector(),pHG.product());
-      }
-      else 
-      {
-	 edm::LogWarning("CaloGeometryBuilder") 
-	    << "Reconstruction geometry requested for a not implemented sub-detector: " 
-	    << (*ite); 
-      }
+   // look for HCAL parts
+   // assume 'HCAL' for all of HCAL.
+   if(hcalToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building HCAL reconstruction geometry";
+
+     auto const& pG = iRecord.get(hcalToken_);
+     pCalo->setSubdetGeometry( DetId::Hcal, HcalBarrel , &pG );
+     pCalo->setSubdetGeometry( DetId::Hcal, HcalEndcap , &pG );
+     pCalo->setSubdetGeometry( DetId::Hcal, HcalOuter  , &pG );
+     pCalo->setSubdetGeometry( DetId::Hcal, HcalForward, &pG );
+   }
+   if(zdcToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building ZDC reconstruction geometry";
+     auto const& pG = iRecord.get( zdcToken_ );
+     pCalo->setSubdetGeometry( DetId::Calo, HcalZDCDetId::SubdetectorId, &pG);
+   }
+   if(castorToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building CASTOR reconstruction geometry";
+     auto const& pG = iRecord.get( castorToken_ );
+     pCalo->setSubdetGeometry( DetId::Calo, HcalCastorDetId::SubdetectorId, &pG);
+   }
+
+   // look for Ecal Barrel
+   if(ecalBarrelToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building EcalBarrel reconstruction geometry";
+     auto const& pG = iRecord.get( ecalBarrelToken_ );
+     pCalo->setSubdetGeometry(DetId::Ecal, EcalBarrel, &pG);
+   }
+   // look for Ecal Endcap
+   if(ecalEndcapToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building EcalEndcap reconstruction geometry";
+     auto const& pG = iRecord.get( ecalEndcapToken_ );
+     pCalo->setSubdetGeometry(DetId::Ecal, EcalEndcap, &pG);
+   }
+   // look for Ecal Preshower
+   if(ecalPreshowerToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building EcalPreshower reconstruction geometry";
+     const auto& pG = iRecord.get(ecalPreshowerToken_);
+     pCalo->setSubdetGeometry(DetId::Ecal, EcalPreshower, &pG);
+   }
+
+   // look for TOWER parts
+   if(caloTowerToken_.isInitialized()) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building TOWER reconstruction geometry";
+     const auto& pG = iRecord.get(caloTowerToken_);
+     pCalo->setSubdetGeometry(DetId::Calo, 1, &pG);
    }
+
+   for(auto const& hgcTokenLabel: hgcalTokens_) {
+     edm::LogInfo("CaloGeometryBuilder") << "Building " << hgcTokenLabel.second << " reconstruction geometry";
+     auto const& pHG = iRecord.get(hgcTokenLabel.first);
```

Comment:
@Dr15Jones Should we consider a mechanism similar to `labelsForToken(EDGetToken, ProductLabels&)` to get the ESProduct label(s) based on the token?

Reply:
That is on the 'to do' list.